### PR TITLE
Feature/optional wavelets normalization

### PR DESCRIPTION
**Pull request by lucaf**
This PR:
 - Adds an optional flag to generate normalized (with unit energy) Morlet wavelets & updates the GUI accordingly
 - Adds an optional parameter to specify the wavelets cycles & updates the GUI accordingly
 - Updates the code that manages the wavelet transform configuration file, where the new flag is added 
 - Fixes the root path of the project in the code, accordingly to the new root: WTools

### CODE REVIEW

File: sh/check_evok_log.m [hunk 1]
```diff
@@ -17,15 +17,21 @@
     tf_cmor_cfg;
     
     %GET log value before averaging
-    if length(defaultanswer)>=12
-        varargout{1}=defaultanswer{1,11};
+    if length(defaultanswer)>=10
```
Comment (lucaf) [inline sh/check_evok_log.m]: @eugenioparise Here I've changed the index for the 'log' flag as 11 looked wrong: to be confirmed. 
Secondly: do not understand the logic of the >=, like the files wasn't fixed length. What's the reason for it? 
I've followed the same logic to set the second return value (evok flag), but it doesn't make sense to me. It probably needs to be changed: if the cfg file doesn't store 12 values, then an error should be returned.

**Comment by lucaf** [comment]
@ambrafer @eugenioparise Pls review the last changes and test that everything works as expected. Unfortunately  I couldn't test the whole flow with the data set provided by Eugenio, for I get an error at some point, related to some matrix dimension mismatch. However Eugenio, you seemed to make it working from the beginning to the end in your previous test. So I'm confident. @ambrafer: if there are no issues, I think we can merge the PR.

File: sh/tf_cmor.m
Comment (eugenioparise) [inline sh/tf_cmor.m:1]: @lucaf I just tried to process a dataset and I am getting the following error:

Undefined function or variable 'fb'.

Error in tf_cmor (line 510)
        average(EEG,OutPath,Fa,tmin,tmax,tres,'cwt',fb,ChannelsList,[0  0
        0  1  0
        0],0,EpochsList,cwtmatrix,extraedges,logtransform,varargin);

Error in wtools>WTPushButton_Callback (line 273)
tf_cmor();

Error in gui_mainfcn (line 95)
        feval(varargin{:});

Error in wtools (line 60)
    gui_mainfcn(gui_State, varargin{:});

Error in
@(hObject,eventdata)wtools('WTPushButton_Callback',hObject,eventdata,guidata(hObject))

 
Error while evaluating UIControl Callback

**Review by eugenioparise — CHANGES_REQUESTED** [review]

File: sh/tf_cmor.m
Comment (eugenioparise) [reply sh/tf_cmor.m:1]: @lucaf tf_cmor.m line 510 replace fb with n? Is it correct? I tried locally on my laptop and it did work, but please doublechcek. Thanks!

Comment (lucaf) [reply sh/tf_cmor.m:1]: @eugenioparise  Ahi... I missed that fb reference: no it should be n/2. Let me fix it....

**Comment by lucaf** [comment]
Should be fxed

**Comment by eugenioparise** [comment]
@lucaf So....

Lines 511 and 517 tf_cmor.m: text messages should be 'TEXT'  and not "TEXT".

Second, I am getting an error message that I have never seen before, I am attaching a screenshot.

Screenshot 2023-08-05 at 11 28 33

To be more clear: I fixed the "" to '' locally on my laptop, and when clicking Ok on the tf_cmor.m GUI I get the error I mentioned.

**Comment by lucaf** [comment]
@eugenioparise  matlab doesn't seem to be bothered if I you use " " for the fprintf format string. If you get bad min time value is because I added a test on time value to be >= 0, as it seemed reasonable to me, unless min time can be indeed negative. Is that the case?

**Comment by eugenioparise** [comment]
@lucaf My version of Matlab, 2014a, doesn't like " ". Perhaps for back compatibility is better to use ' ' all the time.
I saw the time test you implemented, thanks! However, -500 is actually the correct min value: the segments in the dataset I am using (and almost in all EEG datasets) are counted from the stimulus onset and therefore go from a negative value at the left edge to a positive value at the right edge. The segments in my dataset are from -500 to 1500.

**Comment by lucaf** [comment]
@eugenioparise  tmin check fixed, double quotes replaced with single quotes in the fprintf

**Comment by lucaf** [comment]
@eugenioparise With the last update I tried to improve the information displayed on the transform parameters input window and the checks on the value of the parameters. There's one thing which I discovered: when "compute evoked oscillations" is checked, EEG.data=mean(EEG.data,3) is performed, which actually change the  number of epochs so that the user defined list of epochs to process depends actually on that option: if set, the list of epochs must be [] or [1].
Could you pls double check?

Also, could you run this test, in which first you select all the channels, then just a subset, then again all the channels. Are the pllots for all the 3 cases what you'd expect to see (I mean in terms of number)? Thx

**Comment by ambrafer** [comment]
@lucaf when I specify the time limits for the wavelet transform, it doesn't let me specify the highest value as the right limit. See the example below: I have a time series from -400 to 998 ms; accordingly, I want to specify -400 as the limit on the left-hand side and 998 as the limit on the right-hand side.  
Capture

However, it then says that 998 is not accepted:
Capture2

**Comment by ambrafer** [comment]
@lucaf note that it works if I put 997 or below...
also, I get strange results with several epochs that do not contain any data points. See a screenshot below: you notice many instances with "remaining 0 minutes". It's like this for all subjects and conditions and throughout the epochs.
Capture3

**Comment by lucaf** [comment]
> @lucaf when I specify the time limits for the wavelet transform, it doesn't let me specify the highest value as the right limit. See the example below: I have a time series from -400 to 998 ms; accordingly, I want to specify -400 as the limit on the left-hand side and 998 as the limit on the right-hand side. Capture
> 
> However, it then says that 998 is not accepted: Capture2

@ambrafer Might have to do with rounding....  Also the upper bound range should not be displayed with scientific notation... Need to check.

**Comment by lucaf** [comment]
> @lucaf note that it works if I put 997 or below... also, I get strange results with several epochs that do not contain any data points. See a screenshot below: you notice many instances with "remaining 0 minutes". It's like this for all subjects and conditions and throughout the epochs. Capture3

@ambrafer Why strange results? What is strange exactly? The epochs come from the EEG struct which is read directly from the file.

**Comment by ambrafer** [comment]
> > @lucaf note that it works if I put 997 or below... also, I get strange results with several epochs that do not contain any data points. See a screenshot below: you notice many instances with "remaining 0 minutes". It's like this for all subjects and conditions and throughout the epochs. Capture3
> 
> @ambrafer Why strange results? What is strange exactly? The epochs come from the EEG struct which is read directly from the file.

the odd bit is that I get several epochs that do not contain any data points: in the screenshot, you notice many instances with "remaining 0 minutes". @eugenioparise do you get the same problem?

**Comment by eugenioparise** [comment]
@ambrafer @lucaf I remember that WTools has a piece of code in average.m that cuts out epochs with 0 amplitude, that is flat epochs. I do not remember whether I have observed that problem ("0 minutes"), but it might simply be the effect of average.m cutting out flat epochs. If it's just this, you should get a message on Matlab console telling you the number of flat epochs that were cut.

**Comment by ambrafer** [comment]
> @ambrafer @lucaf I remember that WTools has a piece of code in average.m that cuts out epochs with 0 amplitude, that is flat epochs. I do not remember whether I have observed that problem ("0 minutes"), but it might simply be the effect of average.m cutting out flat epochs. If it's just this, you should get a message on Matlab console telling you the number of flat epochs that were cut.

@eugenioparise @lucaf it's true, I get the warning (see below):

Capture4

And according to this warning, I now realize that it treats all epochs as flat (here 167.167)! Same happens for all conditions and subjects.
In fact later on, when I try to do chopping and baseline, it crashes because (I believe) it cannot find any non-flat epochs to work with (see below):

Capture5

**Comment by ambrafer** [comment]
> > @ambrafer @lucaf I remember that WTools has a piece of code in average.m that cuts out epochs with 0 amplitude, that is flat epochs. I do not remember whether I have observed that problem ("0 minutes"), but it might simply be the effect of average.m cutting out flat epochs. If it's just this, you should get a message on Matlab console telling you the number of flat epochs that were cut.
> 
> @eugenioparise @lucaf it's true, I get the warning (see below):
> 
> Capture4
> 
> And according to this warning, I now realize that it treats all epochs as flat (here 167.167)! Same happens for all conditions and subjects. In fact later on, when I try to do chopping and baseline, it crashes because (I believe) it cannot find any non-flat epochs to work with (see below):
> 
> Capture5

In general, something which is not clear to me is that we compute the wavelet in **tf_mor**, but then we call **average**, which in turn calls **complexwt**, which in turn again computes the wavelet?? it's not clear to me what average and complexwt do, and if there is currently a conflict with tf_mor.

**Comment by eugenioparise** [comment]
@ambrafer @lucaf So… if all epochs are treated as flat is probably because all epochs ARE flat. This means very likely that the datasets were not imported in the right way. Try to load the dataset you are processing in EEGLAB and plot the waves: you will see that something is wrong.

tf_cmor.m prepares everything and computes the wavelets; average.m runs the average of transformed epochs (or of the ERP if evok is selected); complexwt.m does the transformation, that is the convolution of the epochs with the wavelets. That are not in conflict, they work together… like the 3 of us 😊
Clear?

**Comment by ambrafer** [comment]
> @ambrafer @lucaf So… if all epochs are treated as flat is probably because all epochs ARE flat. This means very likely that the datasets were not imported in the right way. Try to load the dataset you are processing in EEGLAB and plot the waves: you will see that something is wrong.
> 
> tf_cmor.m prepares everything and computes the wavelets; average.m runs the average of transformed epochs (or of the ERP if evok is selected); complexwt.m does the transformation, that is the convolution of the epochs with the wavelets. That are not in conflict, they work together… like the 3 of us 😊 Clear?

@eugenioparise thanks for the clarification. However, I believe the epochs are imported correctly: when I plot the data in EEGLAB the waves look allright, and the TF decomposition works both in EEGLAB and in WTools (using the original version). So I still believe that there is something not working properly in the new version of WTools, at least when **I** use it. Everything works allright for you?

**Comment by eugenioparise** [comment]
@ambrafer Yesterday the new version was working perfectly on my laptop.
Maybe @lucaf should have a look. But meanwhile you could try with the three datasets I shared for the figures. This is quite strange…

**Comment by eugenioparise** [comment]
@ambrafer Yesterday the new version was working perfectly on my laptop.
Maybe @lucaf should have a look. But meanwhile you could try with the three datasets I shared for the figures. This is quite strange…

**Comment by lucaf** [comment]
> @ambrafer Yesterday the new version was working perfectly on my laptop. Maybe @lucaf should have a look. But meanwhile you could try with the three datasets I shared for the figures. This is quite strange…

@ambrafer My epochs are fine. Pls check the content of your EEG struct in tf_cmor.m, just after it's loaded (debug, BP @line 164, EEG.data). Also I do not get the error on the tmax:  I didn't get it before (I mean with the version you were using) and of course I do not get it after the last commit where I hope to have fixed the type of the timeRange var (which now shouldn't be displayed in scientific notation any more)
image

**Comment by ambrafer** [comment]
> > @ambrafer Yesterday the new version was working perfectly on my laptop. Maybe @lucaf should have a look. But meanwhile you could try with the three datasets I shared for the figures. This is quite strange…
> 
> @ambrafer My epochs are fine. Pls check the content of your EEG struct in tf_cmor.m, just after it's loaded (debug, BP @line 164, EEG.data). Also I do not get the error on the tmax: I didn't get it before (I mean with the version you were using) and of course I do not get it after the last commit where I hope to have fixed the type of the timeRange var (which now shouldn't be displayed in scientific notation any more) image

I pulled your latest update. Both errors are gone now: 1. error on the tmax (visualization in the GUI also fixed, no scientific notation anymore) and 2. issue with the wavelet transformation / flat epochs.
I used the same datasets and the same import settings as before.

**Comment by ambrafer** [comment]
> @eugenioparise With the last update I tried to improve the information displayed on the transform parameters input window and the checks on the value of the parameters. There's one thing which I discovered: when "compute evoked oscillations" is checked, EEG.data=mean(EEG.data,3) is performed, which actually change the number of epochs so that the user defined list of epochs to process depends actually on that option: if set, the list of epochs must be [] or [1]. Could you pls double check?
> 
> Also, could you run this test, in which first you select all the channels, then just a subset, then again all the channels. Are the pllots for all the 3 cases what you'd expect to see (I mean in terms of number)? Thx

@eugenioparise there is still this open issue

**Comment by eugenioparise** [comment]
@ambrafer I did reply to that post already... but I can't see my answer though...
@lucaf did you actually get my reply?

**Comment by lucaf** [comment]
> @ambrafer I did reply to that post already... but I can't see my answer though... @lucaf did you actually get my reply?

I think I've lost track: github has an horrible interface for messaging. What was it?

**Comment by eugenioparise** [comment]
@lucaf @ambrafer So... I tried everything and all works fine. About the evoked oscillation option: it is normal that the number of epochs are reduced to just one, because evoked oscillations are computed on the ERP (the average of all epochs).
I do think that putting the max number of epoch on tf_cmor.m GUI is misleading. It is a not predictable and not consistent number (each participant will have a different number of epochs for each condition). It is perfectly fine for the number of electrodes, because that MUST be consistent across participants and across condition for the whole experiment.

I also tried the plotting you suggested (all; subset; all) and everything worked well (using X Average.m).

**Comment by lucaf** [comment]
> @lucaf @ambrafer So... I tried everything and all works fine. About the evoked oscillation option: it is normal that the number of epochs are reduced to just one, because evoked oscillations are computed on the ERP (the average of all epochs). I do think that putting the max number of epoch on tf_cmor.m GUI is misleading. It is a not predictable and not consistent number (each participant will have a different number of epochs for each condition). It is perfectly fine for the number of electrodes, because that MUST be consistent across participants and across condition for the whole experiment.
> 
> I also tried the plotting you suggested (all; subset; all) and everything worked well (using X Average.m).

@eugenioparise Ok. Then I do not understand why there's an option at all to specify the epochs, which is shared  across subjects, if each subject may have a different number of epochs... That is a bit misleading as well IMO or at least not entirely clear. The best would be to read all the subjects data and find a max number for the epochs. Actually, it'd be advisable to have single data structure that stores all the basic measurements info about each subject, loaded once, at the beginning of the analysis. I won't do that now though.... I will change the test on the epochs such that they have to be >= 1 and adapt the set when processing each subject.

**Comment by eugenioparise** [comment]
@lucaf Yes, I agree with you! That option is there because if the user wants (for whatever reason) process only the first n epochs (for every condition for each participant), they can do it. Of course the assumption is that there are always n epochs to process in that particular project. So, it's necessary that the user knows this information, and that they know what they are doing.

**Comment by ambrafer** [comment]
> @lucaf Yes, I agree with you! That option is there because if the user wants (for whatever reason) process only the first n epochs (for every condition for each participant), they can do it. Of course the assumption is that there are always n epochs to process in that particular project. So, it's necessary that the user knows this information, and that they know what they are doing.

@eugenioparise @lucaf do we want to give visibility to this feature (i.e. the possibility to process only n epochs)? do we 1. comment the section accordingly (advising the user to be careful etc...) 2. add an option in the GUI or 3. even add a description in the tutorial or paper? Maybe 1. is enough? From what you say, it seems that it's rather rare that people want to process only a subset of the epochs... what do you think?

**Comment by lucaf** [comment]
> > @lucaf Yes, I agree with you! That option is there because if the user wants (for whatever reason) process only the first n epochs (for every condition for each participant), they can do it. Of course the assumption is that there are always n epochs to process in that particular project. So, it's necessary that the user knows this information, and that they know what they are doing.
> 
> @eugenioparise @lucaf do we want to give visibility to this feature (i.e. the possibility to process only n epochs)? do we 1. comment the section accordingly (advising the user to be careful etc...) 2. add an option in the GUI or 3. even add a description in the tutorial or paper? Maybe 1. is enough? From what you say, it seems that it's rather rare that people want to process only a subset of the epochs... what do you think?

I think we should keep it. Just need to be more clear about how that settings will be used in the processing. On that: I do not understand what those epochs are :-)  Could you explain pls? Are epoch set of contiguous integer from 1 to maxEpochs or can be that a subject misses say epoch 1, 3, 6 but has 2 4 5...?

**Comment by eugenioparise** [comment]
@lucaf @ambrafer  I agree, we should keep it, but we should be more clear on what it means/does. Probably the best is to explain it well in the tutorial (and honestly it's nothing very complicated to explain).

@lucaf EEG epochs are segments of EEG cut - from the continuous EEG - around an event (typically the stimulus onset). So, they are arrays of floating point numbers. Of course each epoch will  be made of n array, where n is the number of electrodes on the head of the participant. Epochs goes from 1 to maxEpochs (= the max number of administered stimuli in one condition), but it's unlikely that all epochs are good (= artefact free). As a consequence, the experimenter must clean the EEG, throwing away contaminated epochs. So, it's perfectly normal that a subject misses say epoch 1, 3, 6 but has 2 4 5... in one condition; misses epoch 1, 4, 5 but has 2, 3, 6 in another condition. Also also the final number of accepted epochs in two conditions can be slightly different, e.g 36 in condition A and 29 in condition B.
However, I don't think we need to explain these things in the tutorial: they are pretty much obvious, even for beginners. The details of the exact processing pipeline we use may not be obvious though.

**Comment by lucaf** [comment]
@eugenioparise Thx for the explanation.

**Comment by lucaf** [comment]
Guys can I close this PR? Is there anything else to do?

**Comment by eugenioparise** [comment]
@lucaf As far as I can tell you can close it. I am sure that when people will start using it other things will come out :-)

**Pull request merged**